streaming/base/dataset.py

```diff
@@ -367,201 +456,166 @@ def _get_progress(self, world: World) -> Tuple[int, int]:
+            name = f'{self._prefix}_epoch_shape'
+            size = 5 * np.int64().nbytes
+            shape_shm = SharedMemory(name, True, size)
+            shape_shm.buf[:size] = np.array(sample_ids.shape, np.int64).tobytes()
+
+            # Save the generated epoch data to shared memory.
+            name = f'{self._prefix}_epoch_data'
+            size = sample_ids.size * np.int64().nbytes
+            data_shm = SharedMemory(name, True, size)
+            data_shm.buf[:size] = sample_ids.tobytes()
+
+            # After barrier: all other workers on the same node load the sample ID tensor.
+            self._worker_barrier(world.workers_per_node)
+        else:
+            # Before barrier: local rank zero generates the sample ID tensor.
+            self._worker_barrier(world.workers_per_node)
 
-        # Is compression used?
-        if zip_info:
-            # Download the compressed form if missing.
-            zip_filename = os.path.join(self.local, self.split, zip_info.basename)
-            if not os.path.isfile(zip_filename):
-                self._download_file(zip_info.basename)
+            # Load the generated epoch shape from shared memory.
+            name = f'{self._prefix}_epoch_shape'
+            size = 5 * np.int64().nbytes
+            shape_shm = SharedMemory(name, False, size)
+            shape = tuple(np.ndarray(5, buffer=shape_shm.buf, dtype=np.int64))
 
-            # Validate and decompress.
-            self._decompress_shard_part(zip_info, zip_filename, raw_filename, compression)
-        else:
-            # Download the raw form.
-            self._download_file(raw_info.basename)
+            # Attach to the generated epoch data in shared memory.
+            name = f'{self._prefix}_epoch_data'
+            size = int(np.prod(shape)) * np.int64().nbytes
+            data_shm = SharedMemory(name, False, size)
+            sample_ids = np.ndarray(shape, buffer=data_shm.buf, dtype=np.int64)
 
-            # Validate if requested.
-            if self.validate_hash:
-                data = open(raw_filename, 'rb').read()
-                if get_hash(self.validate_hash, data) != raw_info.hashes[self.validate_hash]:
-                    raise ValueError(f'Checksum failure: {raw_filename}')
+        # Each worker gets their portion of the work.
+        sample_ids = sample_ids[world.node, world.rank_of_node, world.worker_of_rank].flatten()
 
-    def _download_shard(self, shard_id: int) -> None:
-        """Download the given shard.
+        # Wait for everyone to get their part.
+        self._worker_barrier(world.workers_per_node)
 
-        Args:
-            shard_id (int): Shard ID.
-        """
-        reader = self.shards[shard_id]
-        for raw_info, zip_info in reader.file_pairs:
-            self._download_shard_part(raw_info, zip_info, reader.compression)
+        # Now clean up after ourselves.
+        if world.is_local_leader:
+            shape_shm.unlink()
```

Comment:
I am afraid `unlink()` method gets called by `num_workers` processes from dataloader and `unlink()` should only be called once per processes, otherwise it will throw some nasty error. Should the cleanup thing move as part of `__del__()` method ??

Also, you might want to have `close()` method first and then `unlink()` for proper cleanup.